### PR TITLE
Refactor utils to support Agent resource

### PR DESCRIPTION
**Pull request by falkerson**
This change is [Reviewable](https://reviewable.io/reviews/mirantis/k8s-netchecker-server/94)

### CODE REVIEW

File: pkg/extensions/controller/controller.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+		return
+	}
+
+	agentCopy := copyObj.(*ext_v1.Agent)
+	agentCopy.Status = ext_v1.AgentStatus{
+		State:   ext_v1.AgentStateProcessed,
+		Message: "Successfully processed by controller",
+	}
+
+	err = c.AgentClient.Put().
```
Comment (AlexeyKasatkin) [inline pkg/extensions/controller/controller.go]: Does not it cause side effects (that object is changed as soon as it was added into cache)?

File: pkg/utils/handler.go [hunk 2]
```diff
@@ -23,32 +24,71 @@ import (
...
+		err = ext_client.CreateAgentThirdPartyResource(clientset)
+		if err != nil && !api_errors.IsAlreadyExists(err) {
+			return nil, err
+		}
+
+		ext, err := ext_client.WrapClientsetWithExtensions(clientset, config)
+		if err != nil {
+			return nil, err
+		}
+		// wait until TPR gets processed
```
Comment (AlexeyKasatkin) [inline pkg/utils/handler.go]: ... gets created?

File: pkg/extensions/client/agent.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	_, err := clientset.ExtensionsV1beta1().
+		ThirdPartyResources().
+		Create(agent)
+
+	return err
+}
+
+// WaitForAgentResource is a handler to check 3rd-party resource exist
+func WaitForAgentResource(client *rest.RESTClient) error {
+	return wait.Poll(100*time.Millisecond, 60*time.Second, func() (bool, error) {
```
Comment (AlexeyKasatkin) [inline pkg/extensions/client/agent.go]: Is not 60 seconds too long? Maybe make it a variable?

File: pkg/utils/handler.go [hunk 4]
```diff
@@ -79,36 +120,88 @@ func (h *Handler) AddMiddleware() {
...
+		Status: ext_v1.AgentStatus{
+			State:   ext_v1.AgentStateCreated,
+			Message: "Created, not processed yet",
+		},
+	}
+
+	agent, err = h.ExtensionsClientset.Agents().Create(agent)
+
+	if err != nil {
+		glog.V(5).Info(err)
```
Comment (AlexeyKasatkin) [inline pkg/utils/handler.go]: why informational?

File: pkg/utils/handler.go [hunk 5]
```diff
@@ -79,36 +120,88 @@ func (h *Handler) AddMiddleware() {
...
+	}
+
+	err = ext_client.WaitForAgentInstanceProcessed(h.ExtensionsClientset, agentName)
+	if err != nil {
+		glog.V(5).Info(err)
```
Comment (AlexeyKasatkin) [inline pkg/utils/handler.go]: it is better to be an error message, afaic.

File: pkg/utils/handler.go [hunk 6]
```diff
@@ -79,36 +120,88 @@ func (h *Handler) AddMiddleware() {
...
 	agentData.LastUpdated = time.Now()
-	glog.V(10).Infof("Updating the agents cache with value: %v", agentData)
+	glog.V(10).Infof("Updating the agents resource with value: %v", agentData)
+
 	agentName := rp.ByName("name")
-	if _, exists := h.AgentCache[agentName]; !exists {
+	_, err := h.ExtensionsClientset.Agents().Get(agentName)
+
+	if err != nil {
+		glog.V(5).Info(err)
```
Comment (AlexeyKasatkin) [inline pkg/utils/handler.go]: should be error message, afaic

File: pkg/utils/handler.go [hunk 7]
```diff
@@ -79,36 +120,88 @@ func (h *Handler) AddMiddleware() {
...
+		},
+	}
+
+	agent, err = h.ExtensionsClientset.Agents().Create(agent)
+
+	if err != nil {
+		glog.V(5).Info(err)
+	}
+
+	err = ext_client.WaitForAgentInstanceProcessed(h.ExtensionsClientset, agentName)
```
Comment (AlexeyKasatkin) [inline pkg/utils/handler.go]: I'm not sure it is a good idea to wait synchronously for such stuff. Seems, we need queue here and in onAdd().

File: pkg/utils/handler.go [hunk 8]
```diff
@@ -79,36 +120,88 @@ func (h *Handler) AddMiddleware() {
...
+	}
 }
 
 func (h *Handler) GetAgents(rw http.ResponseWriter, r *http.Request, _ httprouter.Params) {
-	ProcessResponse(rw, h.AgentCache)
+	agentsData := map[string]ext_v1.AgentSpec{}
+	agents, err := h.ExtensionsClientset.Agents().List()
+
+	if err != nil {
+		glog.V(5).Info(err)
```
Comment (AlexeyKasatkin) [inline pkg/utils/handler.go]: let's check all such places! seems, most of them should be error reports.

File: pkg/utils/handler.go [hunk 9]
```diff
@@ -79,36 +120,88 @@ func (h *Handler) AddMiddleware() {
...
+	}
+
+	for _, agent := range agents.Items {
+		agentsData[agent.ObjectMeta.Name] = agent.Spec
+	}
+
+	if err = ProcessResponse(rw, agentsData); err != nil {
+		return
+	}
```
Comment (AlexeyKasatkin) [inline pkg/utils/handler.go]: we've lost the error here

File: pkg/utils/handler.go [hunk 10]
```diff
@@ -79,36 +120,88 @@ func (h *Handler) AddMiddleware() {
...
+	if api_errors.IsNotFound(err) {
+		glog.V(5).Infof("Agent with name %v is not found in the cache", agentName)
 		http.Error(rw, "There is no such entry in the agent cache", http.StatusNotFound)
 		return
 	}
-	ProcessResponse(rw, aData)
+
+	if err = ProcessResponse(rw, agent); err != nil {
+		return
+	}
```
Comment (AlexeyKasatkin) [inline pkg/utils/handler.go]: we've lost the error here

File: pkg/utils/handler.go [hunk 11]
```diff
@@ -79,36 +120,88 @@ func (h *Handler) AddMiddleware() {
...
 }
 
 func (h *Handler) GetSingleAgent(rw http.ResponseWriter, r *http.Request, rp httprouter.Params) {
-	aName := rp.ByName("name")
-	aData, exists := h.AgentCache[aName]
-	if !exists {
-		glog.V(5).Infof("Agent with name %v is not found in the cache", aName)
+	agentName := rp.ByName("name")
+	agent, err := h.ExtensionsClientset.Agents().Get(agentName)
```
Comment (AlexeyKasatkin) [inline pkg/utils/handler.go]: why not check cache? Please clarify the purpose of the cache you use.

File: pkg/utils/handler.go [hunk 12]
```diff
@@ -164,14 +257,18 @@ func (h *Handler) CheckAgents() ([]string, []string, error) {
 	}
 	for _, pod := range pods.Items {
 		agentName := pod.ObjectMeta.Name
-		agentData, exists := h.AgentCache[agentName]
-		if !exists {
+		agent, err := h.ExtensionsClientset.Agents().Get(agentName)
```
Comment (AlexeyKasatkin) [inline pkg/utils/handler.go]: Is there a method of getting all the data using one request? Maybe List() is suitable for this?

File: pkg/extensions/client/agent.go [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+			Namespace(api_v1.NamespaceDefault).
+			Resource(ext_v1.AgentResourcePlural).
+			DoRaw()
+
+		if err == nil {
+			return true, nil
+		}
+
+		if api_errors.IsNotFound(err) {
+			return false, nil
```
Comment (falkerson) [inline pkg/extensions/client/agent.go]: why it should be timeout?

File: pkg/extensions/client/agent.go [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+		}
+
+		return false, err
+	})
+}
+
+// WaitForAgentInstanceProcessed is a handler to check instance of 3rd-party resource created
```
Comment (falkerson) [inline pkg/extensions/client/agent.go]: done

File: pkg/extensions/controller/controller.go [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+type AgentController struct {
+	AgentClient *rest.RESTClient
+	AgentScheme *runtime.Scheme
+}
+
+// Run starts an Agent resource controller
+func (c *AgentController) Run(ctx context.Context) error {
+	glog.V(5).Info("Watch Agent objects")
+
+	// Watch Example objects
```
Comment (falkerson) [inline pkg/extensions/controller/controller.go]: typo

File: pkg/utils/handler.go [hunk 16]
```diff
@@ -79,36 +120,93 @@ func (h *Handler) AddMiddleware() {
...
+		agent, err = h.ExtensionsClientset.Agents().Create(agent)
+	} else {
+		agent, err = h.ExtensionsClientset.Agents().Update(agent)
 	}
+
+	if err != nil {
+		glog.Error(err)
+	}
+
+	err = ext_client.WaitForAgentInstanceProcessed(h.ExtensionsClientset, agentName)
```
Comment (AlexeyKasatkin) [inline pkg/utils/handler.go]: we wait for onAdd() only. how about onUpdate() ? is not it the same problem (no immediate update)?

File: pkg/utils/handler.go (hunk 11 above)
Comment (AlexeyKasatkin) [inline pkg/utils/handler.go]: let's have caching as a separate story.

File: pkg/utils/handler.go (hunk 12 above)
Comment (AlexeyKasatkin) [inline pkg/utils/handler.go]: It has a big impact on connectivity check speed (it was about 5 second for 3 nodes on the lab).

File: pkg/utils/handler.go (hunk 9 above)
Comment (AlexeyKasatkin) [inline pkg/utils/handler.go]: this "if" statement does not make sense (in its current form, at least).

File: pkg/utils/handler.go (hunk 12 above)
Comment (AlexeyKasatkin) [inline pkg/utils/handler.go]: Timings seem to be OK for now.

File: pkg/extensions/apis/v1/types.go [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+// AgentState type
+type AgentState string
+
+const (
+	// AgentStateCreated constant
+	AgentStateCreated AgentState = "Created"
+	// AgentStateProcessed constant
+	AgentStateProcessed AgentState = "Processed"
```
Comment (AlexeyKasatkin) [inline pkg/extensions/apis/v1/types.go]: Now, we don't need these states at all.

File: pkg/extensions/client/agent.go [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+		if api_errors.IsNotFound(err) {
+			return false, nil
+		}
+
+		return false, err
+	})
+}
+
+// WaitForAgentInstanceProcessed is a handler to check instance of 3rd-party resource was processed
+func WaitForAgentInstanceProcessed(ext Clientset, name string) error {
```
Comment (AlexeyKasatkin) [inline pkg/extensions/client/agent.go]: It's obsolete now.

File: pkg/utils/handler.go [hunk 19]
```diff
@@ -79,36 +103,82 @@ func (h *Handler) AddMiddleware() {
...
+		agent, err = h.ExtensionsClientset.Agents().Create(agent)
+	} else {
+		agent, err = h.ExtensionsClientset.Agents().Update(agent)
 	}
+
+	if err != nil {
+		glog.Error(err)
+	}
+
+	h.Metrics[agentName] = NewAgentMetrics(&agentData)
```
Comment (AlexeyKasatkin) [inline pkg/utils/handler.go]: why to create all the metrics every time?

File: pkg/extensions/client/agent.go (hunk 3 above)
Comment (AlexeyKasatkin) [inline pkg/extensions/client/agent.go]: Not a big problem though.

File: pkg/utils/handler.go [hunk 20]
```diff
@@ -79,36 +103,78 @@ func (h *Handler) AddMiddleware() {
...
+
+	agent := &ext_v1.Agent{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name: agentName,
+		},
+		Spec: agentData,
+	}
+
+	if api_errors.IsNotFound(err) {
 		h.Metrics[agentName] = NewAgentMetrics(&agentData)
```
Comment (AlexeyKasatkin) [inline pkg/utils/handler.go]: Need to check, whether we have h.Metrics[agentName] already (to not loose metrics after restart of a server).

**Comment by AlexeyKasatkin** [comment]
Reviewed 6 of 11 files at r1, 1 of 1 files at r4, 1 of 3 files at r5, 2 of 3 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 24 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-netchecker-server/94)*

**Comment by AlexeyKasatkin** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 24 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-netchecker-server/94#-:-KmlWgeo5FXobO3w94Vq:bnfp4nl)*

**Pull request merged**